File: cs454/A1/python/main.py
import sys

ans = "("

def trees(n_nodes):
    #define function inside function
    def dfs(level, total):
        global ans
        if total == 0:
            #found one answer
            #print( ans + ")"*(2*int(n_nodes)-len(ans)) )
            print (ans + ")" * (level+1))
            return
        if level == 0:
            #at root, have to move deeper
            ans += '('
            dfs(level+1, total-1)
            ans = ans[0:-1]
        else:
            #have real choice, do one at a time
            ans += '('
            dfs(level+1, total-1)
            ans = ans[0:-1]

            ans += ')'
            dfs(level-1, total)
            ans = ans[0:-1]

    #start recursion from root
    dfs(0, n_nodes-1)

def main():
    print ("How many points do you have ?")
    x = sys.stdin.readline();
    trees(int(x))

if __name__ == "__main__":
    main()

File: cs454/A1/ocaml/main.ml
open String;;

let ans = ref "(";;

(*  Remove last char of a string.  *)
let strip_last_char str = 
    if str = "" then "" else
        String.sub str 0 ((String.length str) - 1) ;;

(*  Shorthand for getting length of a string.  *)
let len x = 
    String.length x ;;

(*  Construct a string with all ')' base on given lengths *)
let right_braces cnt_ans n_nodes = 
    String.make (2*n_nodes - (len cnt_ans)) ')'
;;

let rec dfs level total = match (level,total) with
     (* used all points, found one answer *)
     | (_, 0) -> print_string ( !ans ^ (right_braces !ans (int_of_string Sys.argv.(1)) ) ^ "\n" )
     (* can only move down the tree *)
     | (0, t) when t > 0 -> (ans := !ans ^ "(" ; 
                            dfs (level+1) (total-1);
                            ans := strip_last_char !ans)
     (* have a real choice, deal with one at a time *)
     | _  -> ((ans := !ans ^ "(" ; dfs (level+1) (total-1) ; ans := strip_last_char !ans ) ; 
                ( ans := !ans ^ ")" ; dfs (level-1) (total); ans := strip_last_char !ans ))
;;

let trees n_nodes = 
    dfs 0 (n_nodes-1)
;;

trees (int_of_string Sys.argv.(1)) 
